framework/include/constraints/ADMortarScalarBase.h

```diff
@@ -0,0 +1,88 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "ADMortarConstraint.h"
+
+/**
+ * This Constraint adds standardized methods for assembling to a primary
+ * scalar variable associated with the major variables of the Mortar
+ * Constraint object. Essentially, the entire row of the residual and Jacobian
+ * associated with this scalar variable will also be assembled here
+ * using the loops over mortar segments.
+ * This variable is "scalar_variable" in the input file and "kappa"
+ * within the source code.
+ */
+
+class ADMortarScalarBase : public ADMortarConstraint
+{
+public:
+  static InputParameters validParams();
+
+  ADMortarScalarBase(const InputParameters & parameters);
+
+  // Using declarations necessary to pull in computeResidual with different parameter list and avoid
+  // hidden method warning
+  using ADMortarConstraint::computeResidual;
+
+  // Using declarations necessary to pull in computeJacobian with different parameter list and avoid
+  // hidden method warning
+  using ADMortarConstraint::computeJacobian;
+
+  /**
+   * The scalar variable that this kernel operates on.
+   */
+  const MooseVariableScalar & scalarVariable() const { return *_kappa_var_ptr; }
+
+  /**
+   * Computes _var-residuals as well as _kappa-residual
+   */
+  virtual void computeResidual() override;
+  /**
+   * Computes d-_var-residual / d-_var and d-_var-residual / d-jvar,
+   * as well as d-_kappa-residual / d-_var and d-_kappa-residual / d-jvar
+   */
+  virtual void computeJacobian() override;
+
+protected:
+  /**
+   * Method for computing the scalar part of residual at quadrature points
+   */
+  virtual ADReal computeScalarQpResidual() { return 0; }
```

Comment:
Would it make sense to error instead of returning 0?

Reply:
As I'm putting in the documentation, I am permitting the developer to derive from this and not actually use the scalar variable in their derived class; an example of why to do that is in the tensor mechanics test module. So, I didn't put any pure virtual methods into the class, and want it by default to return 0 so that the global residual isn't affected.
The way the logic of the constructor is set up, the `computeScalarQpResidual` is not called if the parameter "scalar_variable" is not defined or if the parameter "compute_scalar_residuals" is set to false (this parameter is set to true by default, but the variable `_compute_scalar_residuals` is constructed as false if "scalar_variable" is not defined).

So what do you think the error should check for? That the user intended to override the function but did not, and so are getting a zero residual from the base class even though they have attached a scalar variable ? Or to do an assert on one of the logical parameters like whether `_compute_scalar_residuals` is actually true or false?

Okay, so I've thought of the scenario: so if ClassA derives from `MortarScalarBase` and ClassB derives from ClassA, and both ClassA and ClassB are registered Moose objects but only ClassB is intended for scalar variables, then I'm expecting a user to always leave the parameter "scalar_variable" empty or unreferenced when they put ClassA in the input file. Only ClassB should override the method to replace the return 0. Now, if ClassA by accident gets the "scalar_variable" set, then maybe that should be considered an error and the message say you should only attach a scalar if you also reimplement this method, or otherwise you should not attach a scalar.

What do you think about that case and the suggested error? Or should it be a warning?

Reply:
Hmm I think you've thought about this a lot and I'd trust your judgement on what you think is best at this time. We can always evolve this as well if we change our minds

Reply:
@lindsayad, I have added an error statement for this instead of a return 0, and I've added test cases to verify the error message prints in the proper accidental situation.

Thus, all my planned changes are now made in the framework for this PR, so I open it up for other reviewer comments.

Note that this PR will now depend on completing PR https://github.com/idaholab/moose/pull/23260.